Implement Register UI & Functions, Improve Error Handling

Abdlerhmansaeed opened the pull request:
![registerUi](https://github.com/user-attachments/assets/c19c946a-dfb2-4c34-b100-11b1ac39c357)
Implement Register UI & Functions, Improve Error Handling
- Completed the registration feature
- Implemented the full UI for the Register screen
- Added necessary functions for user registration
- Fixed an issue in `ApiClient` related to error messages not showing correctly to the user
- Improved `DioErrorHandler` to properly display error messages
- Enhanced error handling logic for better user experience

Abdlerhmansaeed committed:
Implement Register UI & Functions, Improve Error Handling

- Completed the registration feature
- Implemented the full UI for the Register screen
- Added necessary functions for user registration
- Fixed an issue in `ApiClient` related to error messages not showing correctly to the user
- Improved `DioErrorHandler` to properly display error messages
- Enhanced error handling logic for better user experience

Abdlerhmansaeed committed:
- resolved conflict

john-safwat commented on lib/core/utils/validator.dart:
```diff
@@ -16,12 +17,12 @@ class Validator {
       return 'please enter your phone number';
     } else if (numberRegex.hasMatch(number) == false) {
-      return 'Phone number must start with 01 and be 11 digits';
+      return 'number starts with 01 and has 11 digits';
```
don't use hard coded string in your code

Abdlerhmansaeed replied:
done

MoatazAbdElwhab commented on lib/features/auth/data/datasource/auth_remote_data_source.dart:
```diff
@@ -0,0 +1,8 @@
+import '../../../../core/error_handling/exceptions/api_exception.dart';
+import '../model/signup_request_model.dart';
+
+abstract class AuthRemoteDataSource {
+  Future<Either<ApiException, SignUpResponseModel>> signup(SignUpRequestModel request);
+}
```
return response only

MoatazAbdElwhab commented on lib/features/auth/data/datasource/auth_remote_data_source_impl.dart:
```diff
@@ -0,0 +1,50 @@
+  AuthRemoteDataSourceImpl(this._apiClient);
+
+  @override
+  Future<Either<ApiException, SignUpResponseModel>> signup(
+      SignUpRequestModel request) async {
+    Either<ApiException, Map<String, dynamic>> response =
+        await _apiClient.post(
+      ApiConstants.sinUpEndPoint,
+      data: request.toJson(),
+      requiresToken: false,
+    );
+
+    return response.fold(
+      (error) {
+        return Left(
+          ApiException(
+            message: error.message,
+            statusCode: error.statusCode,
+          ),
+        );
+      },
+      (successData) {
+        try {
+          final signUpResponse = SignUpResponseModel.fromJson(successData);
+          return Right(signUpResponse);
+        } catch (e) {
+          return Left(
+            ApiException(
+              message: e.toString(),
+              statusCode: 500,
+            ),
+          );
+        }
+      },
+    );
+  }
+}
```
remove Either

MoatazAbdElwhab commented on lib/features/auth/data/repo/auth_repo_impl.dart:
```diff
@@ -0,0 +1,24 @@
+  AuthRepoImpl(this._authRemoteDataSource);
+  @override
+  Future<Either<ApiException, SignUpResponseModel>> signup(
+      SignUpRequestModel request) async {
+    var response = await _authRemoteDataSource.signup(request);
+    return response;
+  }
```
add try catch

MoatazAbdElwhab commented on lib/features/auth/domain/repo/auth_repo.dart:
```diff
@@ -0,0 +1,9 @@
+import '../../data/model/signup_response_model.dart';
+
+abstract class AuthRepo {
+  Future<Either<ApiException, SignUpResponseModel>> signup(SignUpRequestModel request);
```
return Future<Either<ApiException, void>> signup

MoatazAbdElwhab commented on lib/features/auth/presentation/cubit/auth_cubit.dart:
```diff
@@ -1,8 +1,71 @@
+    emit(state.copyWith(selectedGender: selectedGender ?? ''));
+  }
+
+  void resetSignUpState() {
+  emit(state.copyWith(signUpState: BaseInitialState()));
+}
```
remove this

MoatazAbdElwhab commented on lib/features/auth/presentation/pages/signup_page.dart:
```diff
@@ -0,0 +1,307 @@
+
+            if (state.signUpState is BaseErrorState) {
+              Navigator.pop(context);
+              showDialog(
+                context: context,
+                builder: (context) {
+                  return AlertDialog(
+                    title: const Center(child: Text('Error')),
+                    content: Text(
+                        (state.signUpState as BaseErrorState).errorMessage),
+                    actions: [
+                      ElevatedButton(
+                        onPressed: () {
+                          Navigator.pop(context);
+                          context.read<AuthCubit>().resetSignUpState();
+                        },
+                        child: const Text('OK'),
+                      ),
+                    ],
+                  );
+                },
+              );
```
use dialog utils class

Abdlerhmansaeed replied:
There is no such file

MoatazAbdElwhab commented on lib/features/auth/presentation/pages/signup_page.dart:
```diff
@@ -0,0 +1,307 @@
+                              decoration: const InputDecoration(
+                                  hintText: 'Enter Password',
+                                  labelText: 'Password',
+                                  errorStyle:
+                                      TextStyle(overflow: TextOverflow.fade)),
```
use our text style

MoatazAbdElwhab commented on lib/features/auth/presentation/pages/signup_page.dart:
```diff
@@ -0,0 +1,307 @@
+                          Expanded(
+                            child: TextFormField(
+                              validator: (value) {
+                                if (value !=
+                                    authCubit.passwordController.text) {
+                                  return 'Password does not match';
+                                } else {
+                                  return null;
+                                }
+                              },
```
add vlidator passwords match

Abdlerhmansaeed committed:
- Resolved The mentioned Comments on pull Request

Abdlerhmansaeed committed:
- improved validation Functions
- add a previous value {+2} for the Api requires in SignUp
- add confirm password method on The validator.dart file
- removed the user name validation function cuz we don't have a userName

MoatazAbdElwhab commented on lib/features/auth/presentation/cubit/auth_cubit.dart:
```diff
@@ -1,8 +1,79 @@
+    }
+  }
+
+  void selectGender(String gender) {
+    selectedGenderNotifier.value = gender;
+    selectedGender = gender;
+  }
```
why use 2 data type
pls use value notifier only

Abdlerhmansaeed replied:
done

MoatazAbdElwhab reviewed (commented):
excellent work

john-safwat reviewed (changes requested):
avoid using hard coded strings in your code

Abdlerhmansaeed committed:
Update Code to use locals Using Easy Localization
& Fixed The variable Issue
& update the dialog of the success to redirect the user to login page automatically

Abdlerhmansaeed committed:
Update Code to use locals Using Easy Localization
& Fixed The variable Issue
& update the dialog of the success to redirect the user to login page automatically

Abdlerhmansaeed committed:
Merge remote-tracking branch 'origin/feature/auth' into feature/auth-Register

# Conflicts:
#	pubspec.yaml
